internal/server/server.go

```diff
@@ -17,209 +17,255 @@ import (
+	server StoppableServer,
+	transportType utils.TransportType,
+	errChan chan<- error,
+) {
+	go func() {
+		slog.InfoContext(ctx, "the MCP server is listening",
+			"server.address", listenAddr,
+			"server.transport", transportType,
+		)
+		if err := server.Start(listenAddr); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			slog.ErrorContext(ctx, fmt.Sprintf("failed to serve MCP server via %s", transportType),
+				"error", err,
 			)
+			errChan <- err
 		}
+	}()
+}
 
-		return authContextFunc(
-			c,
-			r,
-			serveOpts.OauthValidateURL,
-			serveOpts.TrentoURL,
-			serveOpts.TrentoUsername,
-			serveOpts.TrentoPassword,
-		)
-	}
-
-	// Create the server, using custom one to handle mcp auth
-	sseServer := mcpserver.NewSSEServer(mcpSrv, mcpserver.WithSSEContextFunc(authContextFuncWrapper))
-
-	// Run the http server for the mcp in a separate goroutine.
-	slog.InfoContext(ctx, "mcp server via SSE starting", "server.address", listenAddr)
-
-	err := sseServer.Start(listenAddr)
-	if err != nil && !errors.Is(err, http.ErrServerClosed) {
-		slog.ErrorContext(ctx, "failed to serve MCP server via HTTP", "error", err)
-		panic(err)
-	}
-
-	slog.InfoContext(ctx, "mcp server via SSE listening successfully", "server.address", listenAddr)
+// startStreamableHTTPServer initializes and starts a custom streamable HTTP server.
+func startStreamableHTTPServer(
+	ctx context.Context,
+	mcpSrv *mcpserver.MCPServer,
+	listenAddr string,
+	serveOpts *ServeOptions,
+	authContext authContextWrapperFn,
+	errChan chan<- error,
+) (*CustomStreamableHTTPServer, error) {
+	streamableServer := NewCustomStreamableHTTPServer(mcpSrv, "/mcp", authContext, serveOpts)
+	startServer(ctx, listenAddr, streamableServer, utils.TransportStreamable, errChan)
+	return streamableServer, nil
+}
 
-	return nil
+// startSSEServer initializes and starts a Server-Sent Events (SSE) server.
+func startSSEServer(
+	ctx context.Context,
+	mcpSrv *mcpserver.MCPServer,
+	listenAddr string,
+	_ *ServeOptions,
+	authContext authContextWrapperFn,
+	errChan chan<- error,
+) (*mcpserver.SSEServer, error) {
+	sseServer := mcpserver.NewSSEServer(mcpSrv, mcpserver.WithSSEContextFunc(authContext))
+	startServer(ctx, listenAddr, sseServer, utils.TransportSSE, errChan)
+	return sseServer, nil
 }
 
-func waitForShutdownStreamable(ctx context.Context, streamableServer *CustomStreamableHTTPServer) {
+// waitForShutdown, once and interrupt signal is received, it gracefully shuts down the server.
+func waitForShutdown(ctx context.Context, server StoppableServer, serverErrChan <-chan error) error {
 	quit := make(chan os.Signal, 1)
 	signal.Notify(quit, os.Interrupt)
-	<-quit
 
-	slog.InfoContext(ctx, "shutting down mcp server")
+	select {
+	case err := <-serverErrChan:
+		return fmt.Errorf("server error: %w", err)
+	case <-quit:
+		slog.DebugContext(ctx, "interrupt signal received, shutting down")
+	case <-ctx.Done():
+		slog.DebugContext(ctx, "context cancelled, shutting down")
```

Comment:
Mainly for tests. Also might be useful if we, from the main Cobra CLI we want to cancel the context at some point.